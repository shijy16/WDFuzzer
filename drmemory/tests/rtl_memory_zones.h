/* This header contains typedefs and prototypes for some ntdll routines.
 *
 * This type information was originally reverse engineered by the Process Hacker
 * project.  This header only contains typedefs, prototypes, and constants, and
 * thus contains no copyrightable information.
 *
 * Project page: http://processhacker.sourceforge.net/
 * Original source: http://processhacker.svn.sourceforge.net/viewvc/
 *   processhacker/2.x/trunk/phlib/include/ntrtl.h?view=markup
 */

#ifdef RTL_MEMORY_ZONES_H_
# error "Cannot include this file multiple times."
#endif
#define RTL_MEMORY_ZONES_H_

#include <windows.h>

typedef LONG NTSTATUS;

/* Memory zone routines and types.  These seem to implement a straightforward
 * pool allocator on top of NtAllocateVirtualMemory that does not grow.
 * When creating a memory zone, we make a minimum space request, and the size of
 * the bookkeeping data structure is added to it and rounded up to the next page
 * boundary.  We can allocate memory beyond the amount we requested but after we
 * hit the next page boundary it will not allocate more blocks.
 */

typedef struct _RTL_MEMORY_ZONE_SEGMENT
{
    struct _RTL_MEMORY_ZONE_SEGMENT *NextSegment;
    SIZE_T Size;
    PVOID Next;
    PVOID Limit;
} RTL_MEMORY_ZONE_SEGMENT, *PRTL_MEMORY_ZONE_SEGMENT;

typedef struct _RTL_MEMORY_ZONE
{
    RTL_MEMORY_ZONE_SEGMENT Segment;
    RTL_SRWLOCK Lock;
    ULONG LockCount;
    PRTL_MEMORY_ZONE_SEGMENT FirstSegment;
} RTL_MEMORY_ZONE, *PRTL_MEMORY_ZONE;

NTSTATUS
(NTAPI *RtlCreateMemoryZone)(
    __out PVOID *MemoryZone,
    __in SIZE_T InitialSize,
    __reserved ULONG Flags
    );

NTSTATUS
(NTAPI *RtlDestroyMemoryZone)(
    __in __post_invalid PVOID MemoryZone
    );

NTSTATUS
(NTAPI *RtlAllocateMemoryZone)(
    __in PVOID MemoryZone,
    __in SIZE_T BlockSize,
    __out PVOID *Block
    );

NTSTATUS
(NTAPI *RtlResetMemoryZone)(
    __in PVOID MemoryZone
    );

NTSTATUS
(NTAPI *RtlLockMemoryZone)(
    __in PVOID MemoryZone
    );

NTSTATUS
(NTAPI *RtlUnlockMemoryZone)(
    __in PVOID MemoryZone
    );

/* Memory block lookaside routines.
 */

NTSTATUS
(NTAPI *RtlCreateMemoryBlockLookaside)(
    __out PVOID *MemoryBlockLookaside,
    __reserved ULONG Flags,
    __in ULONG InitialSize,
    __in ULONG MinimumBlockSize,
    __in ULONG MaximumBlockSize
    );

NTSTATUS
(NTAPI *RtlDestroyMemoryBlockLookaside)(
    __in PVOID MemoryBlockLookaside
    );

NTSTATUS
(NTAPI *RtlAllocateMemoryBlockLookaside)(
    __in PVOID MemoryBlockLookaside,
    __in ULONG BlockSize,
    __out PVOID *Block
    );

NTSTATUS
(NTAPI *RtlFreeMemoryBlockLookaside)(
    __in PVOID MemoryBlockLookaside,
    __in PVOID Block
    );

NTSTATUS
(NTAPI *RtlExtendMemoryBlockLookaside)(
    __in PVOID MemoryBlockLookaside,
    __in ULONG Increment
    );

NTSTATUS
(NTAPI *RtlResetMemoryBlockLookaside)(
    __in PVOID MemoryBlockLookaside
    );
