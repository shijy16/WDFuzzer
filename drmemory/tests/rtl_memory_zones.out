# **********************************************************
# Copyright (c) 2011 Google, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
begin

Testing Rtl*MemoryZone:
# Don't match, different everytime
#zone: 0xXXXXXX
block 0 offset: 0x0001c
block 1 offset: 0x0101c
block 2 offset: 0x0201c
block 3 offset: 0x0301c
can allocate from padding: 1
cannot allocate beyond padding: 1
sizeof(*zone) == block - zone: 1
Segment.NextSegment == NULL: 1
Segment.Size: 20480
Segment.Next: 0x0101c
Segment.Limit: 0x05000
FirstSegment == zone: 1
# Don't match, different everytime
#heap start: 0xXXXXXX, end: 0xXXXXXX
zone is not in heap: 1

Testing Rtl*MemoryBlockLookaside:
# Don't match, different everytime
#lookaside: 0xXXXXXX
lookaside is not in heap: 1
block 0 is not in heap: 1
block 1 is not in heap: 1
block 2 is not in heap: 1
block 3 is not in heap: 1
# XXX: This is unexpected, I don't understand the Rtl*MemoryBlockLookaside
# routines well enough to explain why the fifth allocation succeeds.
#cannot allocate beyond size: 1
# XXX: This also doesn't make sense.  After getting a successful status code we
# get a clearly unaddressable block location.
#block after reset: 00000004
done
~~Dr.M~~ NO ERRORS FOUND:
~~Dr.M~~       0 unique,     0 total unaddressable access(es)
~~Dr.M~~       0 unique,     0 total uninitialized access(es)
~~Dr.M~~       0 unique,     0 total invalid heap argument(s)
~~Dr.M~~       0 unique,     0 total warning(s)
~~Dr.M~~       0 unique,     0 total,      0 byte(s) of leak(s)
~~Dr.M~~       0 unique,     0 total,      0 byte(s) of possible leak(s)
